---
title: "Full_Parameter_Estimation"
author: "Manon Montyne"
date: "2/26/2022"
output: html_document
---
# Load libraries

```{r}
library(dplyr)
library(stringr)
```


# Load in full dataset

```{r}
data <- read.table("phASER_WASP_GTEx_v8_matrix.gw_phased.txt", sep="\t", header=T)
```

# Create output table

```{r}
output <- data[, 1:4]
```


# Loop over GTEX header names

```{r}
GTEX_names <- colnames(data[, -c(1:4)])
```

# Create list of all possible IDs

```{r}
list_of_IDs <- c("contig", "name", "start", "stop")
for (name in GTEX_names){
  sample <- str_extract(name, "GTEX_.+_")     #Extract sample ID
  sample_ID <- substring(sample,6, nchar(sample)-1)
  if (!(sample_ID %in% list_of_IDs)){
    list_of_IDs <- append(list_of_IDs, sample_ID)
  }
}
```

# Calculate rho for each transcript over all the tissues per sample ID (method of moments)

```{r}
for (ID in list_of_IDs[5:842]){
  sample_data <- select(data,contains(ID))
  
  #Empty vectors for succes and total counts
  succes <- c()
  total <- c()
  quadratic <- c()
    
  #Loop over all data to count number of successes and number of trials for each gene
  for (i in 1:56200){
  Q = 0
  N = 0
  S = 0
  for (j in 1:length(sample_data)){
      entry <- sample_data[i, j]
      count1 <- strsplit(entry,"[|]")[[1]][1]
      count1 <- as.integer(count1)
      count2 <- strsplit(entry,"[|]")[[1]][2]
      count2 <- as.integer(count2)
      S = S + count1
      Q = Q + (count1)^2
      N = N + count1 + count2
    }
  succes <- append(succes, S, after=length(succes))
  total <- append(total, N, after=length(total))
  quadratic <- append(quadratic, Q, after=length(quadratic))
  }
  
  #Add columns with the succeses and total number of trials
  sample_data$Succes <- succes
  sample_data$Total <- total
  sample_data$Quadratic <- quadratic
    
    
  #Estimate alfa and beta parameters by m1 and m2
  sample_data$m1 <- sample_data$Succes/sample_data$Total      #m1
  sample_data$m2 <- sample_data$Quadratic/sample_data$Total   #m2

  sample_data$alfa <- ((sample_data$Total*sample_data$m1) - sample_data$m2)/(sample_data$Total*((sample_data$m2/sample_data$m1) - sample_data$m1 - 1) + sample_data$m1) #alfa

  sample_data$beta <- ((sample_data$Total - sample_data$m1)*(sample_data$Total - (sample_data$m2 /sample_data$m1)))/(sample_data$Total*((sample_data$m2/sample_data$m1) - sample_data$m1 - 1) + sample_data$m1) #beta

  #Estimate overdispersion parameter rho
  sample_data$rho <- 1/(sample_data$alfa + sample_data$beta + 1)
  output <- cbind(output, sample_data$rho)
}
```

# Save as csv file

```{r}
names(output) <- list_of_IDs
write.csv(output, "Rho_full_data.csv")
```